### PR TITLE
Cve 2017 5715 2

### PR DESCRIPTION
**Pull request by jenswi-linaro**
This pull request is done on top of https://github.com/OP-TEE/optee_os/pull/2048 to avoid unnecessary conflicts when finally merging everything. The commit to review here is `
core: arm32: thread: invalidate branch predictor`.

### CODE REVIEW

File: core/arch/arm/kernel/thread_a32.S [hunk 1]
```diff
@@ -748,6 +748,51 @@ END_FUNC thread_unwind_user_mode
...
 UNWIND(	.cantunwind)
+#ifdef CFG_CORE_WORKAROUND_SPECTRE_BP
+	/*
+	 * This depends on SP being 8 byte aligned, that is, the lowest
+	 * three bits in SP are zero.
+	 *
+	 * The idea is to form a specific bit pattern in the lowest three
+	 * bits of SP depending on which entry in the vector we enter via.
+	 * This is done by adding 1 to SP in each entry but the last.
+	 */
```
Comment (jforissier) [inline core/arch/arm/kernel/thread_a32.S]: I think this deserves a few words to explain that the BPIALL has to be executed before *any* branch instruction is executed, hence the slightly convoluted code. I know I could/should have made the same comment for the sm commit ;-)

**Comment by jforissier** [comment]
What about 64-bits?

File: core/arch/arm/kernel/thread_a32.S (hunk 1 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/thread_a32.S]: I'll add something.

**Comment by jenswi-linaro** [comment]
You're right I forgot about the 64-bit stuff. I'll add another patch here taking care of that.

**Comment by jenswi-linaro** [comment]
Hmm, the 64-bit branch predictor invalidation is a bit more complicated. I'll take that in a separate pull request.

**Comment by jenswi-linaro** [comment]
Commit updated, tag applied to `core: arm32: thread: invalidate branch predictor` only.

File: core/arch/arm/include/mm/core_mmu.h [hunk 2]
```diff
@@ -74,6 +77,11 @@
 #define CFG_TEE_RAM_VA_SIZE		CORE_MMU_PGDIR_SIZE
 #endif
 
+#ifdef CFG_WITH_LPAE
+#define CORE_MMU_L1_TBL_OFFSET		(CFG_TEE_CORE_NB_CORE * 4 * 8)
```
Comment (etienne-lms) [inline core/arch/arm/include/mm/core_mmu.h]: ` * 4 * 8` looks like magic, especially compared to how `NUM_L1_ENTRIES` is defined in **core_mmu_lpae.c**. Deserves a bit of explanation: why 4 64bit wide mmu descriptors ?

Regarding the armv7, the comment is straightforward: we need a full L1 table.

File: core/arch/arm/kernel/thread.c [hunk 3]
```diff
@@ -883,6 +888,20 @@ static void init_thread_stacks(void)
...
 #endif /*CFG_WITH_PAGER*/
 
+static void init_user_kcode(void)
+{
+#ifdef CFG_CORE_UNMAP_CORE_AT_EL0
+	vaddr_t v;
+
+	v = (vaddr_t)thread_vect_table;
+	thread_user_kcode_va = ROUNDDOWN(v, CORE_MMU_USER_CODE_SIZE);
+	thread_user_kcode_size = CORE_MMU_USER_CODE_SIZE * 2;
```
Comment (etienne-lms) [inline core/arch/arm/kernel/thread.c]: why 2 regions (likely pages)?

File: core/arch/arm/include/mm/core_mmu.h (hunk 2 above)
Comment (jenswi-linaro) [reply core/arch/arm/include/mm/core_mmu.h]: Please make these comments in https://github.com/OP-TEE/optee_os/pull/2048 instead

File: core/arch/arm/kernel/thread.c (hunk 3 above)
Comment (jenswi-linaro) [reply core/arch/arm/kernel/thread.c]: Please make these comments in https://github.com/OP-TEE/optee_os/pull/2048 instead

File: core/arch/arm/include/mm/core_mmu.h (hunk 2 above)
Comment (etienne-lms) [reply core/arch/arm/include/mm/core_mmu.h]: ok, sorry.

File: core/arch/arm/kernel/thread_a32.S (hunk 1 above)
Comment (etienne-lms) [reply core/arch/arm/kernel/thread_a32.S]: minor stuff: s/`doesn't`/`does not`/ , s/`it's`/`it is`/.

Sentence seems wrong to me: `It doesn't matter if it's a conditional or an unconditional branch speculation can still occur.`.

Comment (jenswi-linaro) [reply core/arch/arm/kernel/thread_a32.S]: Perhaps it's a missing comma. How about:
*It doesn't matter if it's a conditional or an unconditional branch, speculation can still occur.*

File: core/arch/arm/kernel/thread_a32.S [hunk 4]
```diff
@@ -748,6 +748,56 @@ END_FUNC thread_unwind_user_mode
 FUNC thread_vect_table , :
 UNWIND(	.fnstart)
 UNWIND(	.cantunwind)
+#ifdef CFG_CORE_WORKAROUND_SPECTRE_BP
```
Comment (jenswi-linaro) [inline core/arch/arm/kernel/thread_a32.S]: I'm wondering if we shouldn't use a different config option here as we're protecting against a different attack and also the defense will be quite expensive in AArch64.

Comment (jforissier) [reply core/arch/arm/kernel/thread_a32.S]: Agreed. `CFG_CORE_WORKAROUND_SPECTRE_BP_SEC` maybe? (rationale: still Spectre branch prediction attack but entirely within secure world hence `_SEC`).

Comment (jenswi-linaro) [reply core/arch/arm/kernel/thread_a32.S]: Sounds good

**Comment by jenswi-linaro** [comment]
Commit `core: arm32: thread: invalidate branch predictor` updated and rebased on top of #2048

**Comment by jforissier** [comment]
Fine with me

**Comment by danh-arm** [comment]
We've discovered that the BPIALL workaround not effective in invalidating the branch predictor for Cortex-A15. An alternative workaround is described in the updated TF security advisory:

https://github.com/ARM-software/arm-trusted-firmware/wiki/Arm-Trusted-Firmware-Security-Advisory-TFV-6

That will affect your implementation here. Note we plan to push an example implementation of this in our SP_MIN payload in the coming days.

Also, it would be good if this workaround was not applied to unaffected CPUs (e.g. Cortex-A32, Cortex-A7 and Cortex-A5).

**Comment by jbech-linaro** [comment]
> We've discovered that the BPIALL ...
> ... That will affect your implementation here. Note we plan to push an example implementation of this in our SP_MIN payload in the coming days.

Thanks for letting us know Dan, this means it's worth for us wait a bit before merging this.

> Also, it would be good if this workaround was not applied to unaffected CPUs (e.g. Cortex-A32, Cortex-A7 and Cortex-A5).

Agree!

**Comment by jforissier** [comment]
Sounds like we may have to use runtime detection in order to install the proper exception vector. Otherwise, we can't properly support mixed configurations like big.LITTLE.

**Comment by etienne-lms** [comment]
Can't we do something about it using the **core/arch/arm/cpus/xxx.mk**?
Current status is the patches are default enable but the platform can force their disabling.

**Comment by jenswi-linaro** [comment]
Rebased on master, updated for runtime configuration.

File: core/arch/arm/kernel/thread_a32.S [hunk 5]
```diff
@@ -460,8 +460,44 @@ KEEP_PAGER thread_rpc
...
+	 */
+	read_midr r1
+	ubfx	r2, r1, #MIDR_IMPLEMENTER_SHIFT, #MIDR_IMPLEMENTER_WIDTH
+	cmp	r2, #MIDR_IMPLEMENTER_ARM
+	bne	1f
+
+	ubfx	r2, r1, #MIDR_PRIMARY_PART_NUM_SHIFT, \
+			#MIDR_PRIMARY_PART_NUM_WIDTH
+
+	mov	r3, #CORTEX_A8_PART_NUM
```
Comment (jforissier) [inline core/arch/arm/kernel/thread_a32.S]:
```
core/arch/arm/kernel/thread_a32.S: Assembler messages:
core/arch/arm/kernel/thread_a32.S:477: Error: invalid constant (c08) after fixup
core/arch/arm/kernel/thread_a32.S:479: Error: invalid constant (c09) after fixup
core/arch/arm/kernel/thread_a32.S:481: Error: invalid constant (c0e) after fixup
core/arch/arm/kernel/thread_a32.S:483: Error: invalid constant (d07) after fixup
core/arch/arm/kernel/thread_a32.S:485: Error: invalid constant (d08) after fixup
core/arch/arm/kernel/thread_a32.S:487: Error: invalid constant (d09) after fixup
core/arch/arm/kernel/thread_a32.S:489: Error: invalid constant (d0a) after fixup
core/arch/arm/kernel/thread_a32.S:494: Error: invalid constant (c0f) after fixup
```

`s/mov/movw/` and `s/movne/movwne/` fixes the issues

File: core/arch/arm/kernel/thread_a32.S [hunk 6]
```diff
@@ -460,8 +460,44 @@ KEEP_PAGER thread_rpc
...
+	cmpne	r2, r3
+	ldreq	r0, =exception_vector_bpial
+	beq	2f
+
+	mov	r3, #CORTEX_A15_PART_NUM
+	cmp	r2, r3
+	ldreq	r0, =exception_vector_a15
+	beq	2f
+#endif
+1:	ldr	r0, =thread_vect_table
```
[the remark this answers is not in the document]
Comment (jenswi-linaro) [reply core/arch/arm/kernel/thread_a32.S]: Yes, let's take that cleanup in a later PR.

**Comment by jenswi-linaro** [comment]
Updated as requested, rebased, squashed and tags applied.

**Comment by jenswi-linaro** [comment]
Pushed updates (commit message and a spell error in a comment) for two checkpatch warnings.
One checkpatch warning for a long line remains, I don't want to break the url.

**Comment by jenswi-linaro** [comment]
Please wait with merging this, this PR requires ACTLR[0] set to 1 for Cortex-A15 platforms. I'm preparing another PR to take care of that.

**Comment by jforissier** [comment]
@jenswi-linaro sure

**Comment by jenswi-linaro** [comment]
Needs https://github.com/OP-TEE/optee_os/pull/2071

**Comment by jenswi-linaro** [comment]
Rebased on master

**Pull request merged**